src/compiler/program.ts

```diff
@@ -765,11 +775,42 @@ namespace ts {
             const modifiedSourceFiles: { oldFile: SourceFile, newFile: SourceFile }[] = [];
             oldProgram.structureIsReused = StructureIsReused.Completely;
 
-            for (const oldSourceFile of oldProgram.getSourceFiles()) {
+            const oldSourceFiles = oldProgram.getSourceFiles();
+            for (const oldSourceFile of oldSourceFiles) {
                 const newSourceFile = host.getSourceFileByPath
                     ? host.getSourceFileByPath(oldSourceFile.fileName, oldSourceFile.path, options.target)
                     : host.getSourceFile(oldSourceFile.fileName, options.target);
 
+                const packageId = oldProgram.getPackageIdOfSourceFile(oldSourceFile);
+                if (packageId) sourceFileToPackageId.set(newSourceFile.path, packageId);
+
+                if (oldSourceFile.redirect) {
+                    // We got `newSourceFile` by path, so it is actually for the underlying file.
+                    // This lets us know if the underlying file has changed. If it has we should break the redirect.
+                    const { underlying } = oldSourceFile.redirect;
+                    if (newSourceFile !== underlying) {
+                        // Underlying file has changed. Might not redirect anymore. Must rebuild program.
+                        return oldProgram.structureIsReused = StructureIsReused.Not;
+                    }
+                    else {
+                        filePaths.push(oldSourceFile.path);
+                        newSourceFiles.push(oldSourceFile);
+                        continue;
+                    }
+                }
+                else if (oldProgram.sourceFileIsRedirectedTo(oldSourceFile)) {
+                    // This is similar to the above case. If a redirected-to source file changes, the redirect may be broken.
+                    if (newSourceFile !== oldSourceFile) {
+                        return oldProgram.structureIsReused = StructureIsReused.Not;
+                    }
+                    else {
+                        sourceFileIsRedirectedToSet.set(oldSourceFile.path, true);
+                        filePaths.push(oldSourceFile.path);
+                        newSourceFiles.push(oldSourceFile);
+                        continue;

```

Comment:
i would let this fall through, and then we will add it to filePaths, and newSoruceFiles